File: windowedadaptation.py
class WindowedAdaptation():
    def __init__(self, warmup, windowsize = 25, windowscale = 2):
        self._windowsize = windowsize
        self._windowscale = windowscale
        self._warmup = warmup
        self._closewindow = self._windowsize
        self._idx = 0
        self._closures = []
        self._num_windows = 0
        self._calculate_windows()

    def _calculate_windows(self):
        if self._warmup > self._windowsize:
            for w in range(self._warmup + 1):
                if w == self._closewindow:
                    self._closures.append(w)
                    self._calculate_next_window()
            self._num_windows = len(self._closures)

    def _calculate_next_window(self):
        self._windowsize *= self._windowscale
        nextclosewindow = self._closewindow + self._windowsize
        if self._closewindow + self._windowscale * self._windowsize >= self._warmup:
            self._closewindow = self._warmup
        else:
            self._closewindow = nextclosewindow

    def window_closed(self, m):
        if self._warmup < self._windowsize:
            return False
        closed = m == self._closures[self._idx]
        if closed and self._idx < self._num_windows - 1:
            self._idx += 1
        return closed

    def reset(self):
        self._idx = 0
        self._closures.clear()
        self._num_windows = 0
        self._calculate_windows()

if __name__ == "__main__":
    warmup = 15_000
    iterations = 30_000
    wa = WindowedAdaptation(warmup, windowsize = 50, windowscale = 2)
    closures = wa._closures
    for m in range(iterations):
        if wa.window_closed(m):
            print(m)

    wa.reset()
    print(wa._closures == closures)

File: windowedadaptation.jl
using Test
using Statistics

struct WindowedAdaptation
    windowsize::Int64
    windowscale::Int64
    warmup::Int64
    closewindow::Int64
    idx::Base.RefValue{Int64}
    closures::Vector{Int64}
    numwindows::Int64
end

function WindowedAdaptation(warmup; windowsize = 50, windowscale = 2)
    closures = []
    closewindow = windowsize
    j = 0
    if warmup > windowsize
        for w in 1:warmup
            if w == closewindow
                push!(closures, w)
                windowsize *= windowscale
                nextclosewindow = closewindow + windowsize
                if closewindow + windowscale * windowsize >= warmup
                    closewindow = warmup
                else
                    closewindow = nextclosewindow
                end
            end
        end
        numwindows = length(closures)
    else
        numwindows = 0
    end
    return WindowedAdaptation(
        windowsize,
        windowscale,
        warmup,
        windowsize,
        Ref(1),
        closures,
        numwindows
    )
end

function window_closed(wa::WindowedAdaptation, m)
    if wa.warmup < wa.windowsize
        return false
    end
    closed = m == wa.closures[wa.idx[]]
    if closed && wa.idx[] < wa.numwindows
        wa.idx[] += 1
    end
    return closed
end

if abspath(PROGRAM_FILE) == @__FILE__
    @testset "WindowedAdaptation test" begin
        warmup = 15_000
        iterations = 30_000
        wa = WindowedAdaptation(warmup)
        for m in 1:iterations
            if window_closed(wa, m)
                println(m)
            end
        end
    end
end
